source/grid/grid_out.cc

```diff
@@ -1470,6 +1474,211 @@ GridOut::write_xfig(const Triangulation<2> &tria,
+            {
+              const auto &f = cell->line(l);
+              if (f->manifold_id() != numbers::flat_manifold_id ||
+                  (f->boundary_id() != 0 &&
+                   f->boundary_id() != numbers::internal_face_boundary_id))
+                set_of_pairs.insert({f->boundary_id(), f->manifold_id()});
+            }
+      }
+    all_pairs = {set_of_pairs.begin(), set_of_pairs.end()};
+
+    int entity = 1;
+    for (const auto &p : set_of_pairs)
+      id_pair_to_entity_tag[p] = entity++;
+  }
+
+  const auto n_entity_tags = id_pair_to_entity_tag.size();
+
+  // All elements in the mesh, by entity tag, and by dealii type.
+  std::vector<std::vector<std::vector<std::size_t>>> element_ids(
+    n_entity_tags, std::vector<std::vector<std::size_t>>(8));
+  std::vector<std::vector<std::vector<std::size_t>>> element_nodes(
+    n_entity_tags, std::vector<std::vector<std::size_t>>(8));
+
+  // One elment id counter for all dimensions.
+  std::size_t element_id = 1;
+
+  const auto add_element = [&](const auto &element, const int &entity_tag) {
+    const auto type = element->reference_cell_type();
+
+    Assert(entity_tag > 0, ExcInternalError());
+    // Add all vertex ids. Make sure we renumber to gmsh, and we add 1 to the
+    // global index.
+    for (const auto &v : element->vertex_indices())
+      element_nodes[entity_tag - 1][type].emplace_back(
+        element->vertex_index(dealii_to_gmsh[type][v]) + 1);
+
+    // Save the element id.
+    element_ids[entity_tag - 1][type].emplace_back(element_id);
+    ++element_id;
+  };
+
+  // Will create a separate gmsh entity, only  if it's a cell, or if the
+  // boundary and/or the manifold ids are not the default ones.
+  // In the meanwhile, also store each pair of dimension and entity tag that was
+  // requested.
+  std::set<std::pair<int, int>> dim_entity_tag;
+
+  auto maybe_add_element =
+    [&](const auto &              element,
+        const types::boundary_id &boundary_or_material_id) {
+      const auto struct_dim  = element->structure_dimension;
+      const auto manifold_id = element->manifold_id();
+
+      // Exclude default boundary/manifold id or invalid/flag
+      const bool non_default_boundary_or_material_id =
+        (boundary_or_material_id != 0 &&
+         boundary_or_material_id != numbers::internal_face_boundary_id);
+      const bool non_default_manifold =
+        manifold_id != numbers::flat_manifold_id;
+      if (struct_dim == dim || non_default_boundary_or_material_id ||
+          non_default_manifold)
+        {
+          const auto entity_tag =
+            id_pair_to_entity_tag[{boundary_or_material_id, manifold_id}];
+          add_element(element, entity_tag);
+          dim_entity_tag.insert({struct_dim, entity_tag});
+        }
+    };
+
+  // Loop recursively over all cells, faces, and possibly lines.
+  for (const auto &cell : tria.active_cell_iterators())
+    {
+      maybe_add_element(cell, cell->material_id());
```

Comment:
Where is the recursion? Why aren't level-cells visited? Can gmsh handle hanging nodes?

Reply:
Gmsh can handle hanging nodes in output, but not in input. 

In a previous version I had a recursion, but it was much more complicated. I replaced the recursive version with loops over cells, faces, and lines: the principle is that every element of every structural dimension of the triangulation is visited and output with its material id (if its a cell), boundary id (if it's a face/line) and manifold id.